### DISCUSSION TITLE
DFT data needed for spin energy using deepmd

### DISCUSSION THREAD

**Discussion by siddarthachar** [deepmodeling/deepmd-kit#3023] [General]
Hi,

I am working with a system where spin contributions are important. I just found out about [https://docs.deepmodeling.com/projects/deepmd/en/master/model/train-energy-spin.html](url) and how DPs can be made to predict spin of the system. I went to the examples folder for spin [https://github.com/deepmodeling/deepmd-kit/tree/master/examples/spin](url) and found that the data does not contain information about spin, like magnetic moment for each atom etc. I am not sure what the spin part of the loss function uses to compute the loss. Also, it would be nice to have an example of using VASP to train a model that accounts for spin energy. 

Thanks!

**Comment by yangtengleo** [comment]
For magnetic systems, the DP model represents spin configurations by introducing pseudo-atoms. The types of pseudo-atoms are added in **type.raw**, the coordinates of pseudo-atoms are added in **coord.raw**, and the parameters related to spin are set in **input.json**. The [example for spin](https://github.com/deepmodeling/deepmd-kit/tree/master/examples/spin) is from NiO system, you can see that in [type.raw](https://github.com/deepmodeling/deepmd-kit/blob/master/examples/spin/data/data_0/type.raw), 0 refers to Ni atom, 1 refers to O atom, and 2 refers to pseudo-atom representing the spin of Ni atom. 

The DP model yields the total potential energy including both lattice and spin, so the loss function consists of predicted errors of energy, atomic forces, magnetic torques (virial for optional). This means when you prepare the training data, you need to calculate the magnetic torques applied on each spin and add them to **force.raw**. More details can be found in this [paper](https://arxiv.org/abs/2304.09606).

As we know, the current standard VASP cannot accurately calculate the magnetic torques on spin. For this reason we have developed a new algorithm **DeltaSpin** for constraining spin, more details can refer to this [paper](https://arxiv.org/abs/2208.04551). Due to the licensing restriction, we cannot provide examples here, please contact Prof. **Ben Xu** (bxu@gscaep.ac.cn) if you have requests.

**Reply by siddarthachar** [reply]
Thanks a lot for the response! Yes, I just looked at the *.raw files and found pseudo-atoms in them to represent spin configurations. I really appreciate your explanation of how magnetic torques can be applied to the forces. Sure, I will contact Prof. Ben Xu regarding DeltaSpin.

**Reply by WilliamBlakecc** [reply]
Dear siddarthachar and yangtengleo,
I notice this great disscussion and I want to ask some questions about this. My system is antiferromagetic and I try to train its potentioal. However, its energy predication is far away from y=x curve. I wonder if I do not use DeepSpin? Is DeepSpin to predicate the direction of spin in MD? and will effect energy predict?
